himbaechel/uarch/gowin/gowin.cc

```diff
@@ -7,230 +7,234 @@
-	ctx->cells.at(ctx->id("rst_IBUF_I"))->setAttr(ctx->id("BEL"), std::string("X0Y4/IOBA"));
-	assign_cell_info();
+void GowinImpl::prePlace()
+{
+    ctx->cells.at(ctx->id("leds_OBUF_O"))->setAttr(ctx->id("BEL"), std::string("X0Y14/IOBA"));
+    ctx->cells.at(ctx->id("leds_OBUF_O_1"))->setAttr(ctx->id("BEL"), std::string("X0Y15/IOBB"));
+    ctx->cells.at(ctx->id("leds_OBUF_O_2"))->setAttr(ctx->id("BEL"), std::string("X0Y20/IOBB"));
+    ctx->cells.at(ctx->id("leds_OBUF_O_3"))->setAttr(ctx->id("BEL"), std::string("X0Y21/IOBB"));
+    ctx->cells.at(ctx->id("leds_OBUF_O_4"))->setAttr(ctx->id("BEL"), std::string("X0Y24/IOBB"));
+    ctx->cells.at(ctx->id("leds_OBUF_O_5"))->setAttr(ctx->id("BEL"), std::string("X0Y25/IOBB"));
+    ctx->cells.at(ctx->id("rst_IBUF_I"))->setAttr(ctx->id("BEL"), std::string("X0Y4/IOBA"));
+    assign_cell_info();
 }
 
-void GowinImpl::pack() {
-	// Trim nextpnr IOBs - assume IO buffer insertion has been done in synthesis
-	const pool<CellTypePort> top_ports{
-			CellTypePort(id_IBUF, id_I),
-			CellTypePort(id_OBUF, id_O),
-	};
-	h.remove_nextpnr_iobs(top_ports);
-	// Replace constants with LUTs
-	const dict<IdString, Property> vcc_params;
-	const dict<IdString, Property> gnd_params;
-	h.replace_constants(CellTypePort(id_GOWIN_VCC, id_V), CellTypePort(id_GOWIN_GND, id_G), vcc_params, gnd_params);
-
-	// disconnect the constant LUT inputs
-	mod_lut_inputs();
-
-	// Constrain directly connected LUTs and FFs together to use dedicated resources
-	int lutffs = h.constrain_cell_pairs(
-			pool<CellTypePort>{{id_LUT1, id_F}, {id_LUT2, id_F}, {id_LUT3, id_F}, {id_LUT4, id_F}},
-			pool<CellTypePort>{{id_DFF, id_D}, {id_DFFE, id_D}, {id_DFFN, id_D}, {id_DFFNE, id_D},
-			  {id_DFFS, id_D}, {id_DFFSE, id_D}, {id_DFFNS, id_D}, {id_DFFNSE, id_D},
-			  {id_DFFR, id_D}, {id_DFFRE, id_D}, {id_DFFNR, id_D}, {id_DFFNRE, id_D},
-			  {id_DFFP, id_D}, {id_DFFPE, id_D}, {id_DFFNP, id_D}, {id_DFFNPE, id_D},
-			  {id_DFFC, id_D}, {id_DFFCE, id_D}, {id_DFFNC, id_D}, {id_DFFNCE, id_D}},1);
-	log_info("Constrained %d LUTFF pairs.\n", lutffs);
+void GowinImpl::pack()
+{
+    // Trim nextpnr IOBs - assume IO buffer insertion has been done in synthesis
+    const pool<CellTypePort> top_ports{
+            CellTypePort(id_IBUF, id_I),
+            CellTypePort(id_OBUF, id_O),
+    };
+    h.remove_nextpnr_iobs(top_ports);
+    // Replace constants with LUTs
+    const dict<IdString, Property> vcc_params;
+    const dict<IdString, Property> gnd_params;
+    h.replace_constants(CellTypePort(id_GOWIN_VCC, id_V), CellTypePort(id_GOWIN_GND, id_G), vcc_params, gnd_params);
+
+    // disconnect the constant LUT inputs
+    mod_lut_inputs();
+
+    // Constrain directly connected LUTs and FFs together to use dedicated resources
+    int lutffs = h.constrain_cell_pairs(
+            pool<CellTypePort>{{id_LUT1, id_F}, {id_LUT2, id_F}, {id_LUT3, id_F}, {id_LUT4, id_F}},
```

Comment:
As a style point, probably make these `pool<CellTypePort>` variables rather than having them inline as long initialiser lists in the function parameters.